llmfoundry/models/inference_api_wrapper/openai_causal_lm.py

```diff
@@ -0,0 +1,272 @@
+            assert isinstance(new_batch['continuation_indices'], list)
+            new_batch['continuation_indices'].append(cont_idxs)
+
+        new_batch.update({
+            k: torch.stack(new_batch[k])  # pyright: ignore
+            for k in ['input_ids', 'labels']
+        })
+
+        new_batch.update({k: v for k, v in batch.items() if k not in new_batch})
+
+        return new_batch
+
+    def eval_forward(self, batch: Batch, outputs: Optional[Any] = None):
+        # Override the base class because Chat's API always strips spacing from model outputs resulting in different tokens
+        # than what the continuation would expect.
+        # Get around this issue by retokenizing the batch to remove spacing from the continuation as well as
+        # decoding the whole continuation at once.
+        output_logits_batch = []
+        batch = self.rebatch(batch)
+        for tokens, cont_idxs in zip(batch['input_ids'],
+                                     batch['continuation_indices']):
+
+            seqlen = tokens.shape[0]
+            tokens = tokens.tolist()
+            cont_idxs = cont_idxs.tolist()
+            expected_cont_tokens = tokens[cont_idxs[0]:cont_idxs[-1] + 1]
+            output_logits = torch.nn.functional.one_hot(
+                torch.tensor(tokens[1:cont_idxs[0]]),
+                num_classes=self.tokenizer.vocab_size)
+
+            prompt = self.tokenizer.decode(tokens[:cont_idxs[0]])
+            next_logit_tensor = self.get_next_token_logit_tensor(
+                prompt, num_tokens=len(expected_cont_tokens))
+
+            if next_logit_tensor is not None:
+                output_logits = torch.cat([output_logits, next_logit_tensor])
+            padding = torch.nn.functional.one_hot(
+                torch.full((seqlen - output_logits.shape[0],),
+                           self.tokenizer.pad_token_id),
+                num_classes=self.tokenizer.vocab_size)
+            output_logits = torch.cat([output_logits, padding])
+            output_logits_batch.append(output_logits)
+
+        return torch.stack(output_logits_batch).to(batch['input_ids'].device)
+
+    def get_next_token_logit_tensor(self, prompt: str, num_tokens: int = 1):
+        tries = 0
+        chat_completion = None
+        while tries < MAX_RETRIES:
+            tries += 1
+            try:
+                chat_completion = openai.ChatCompletion.create(
+                    model=self.model_name,
+                    messages=[{
+                        'role': 'user',
+                        'content': prompt
+                    }],
+                    max_tokens=num_tokens,
+                    temperature=0.0)
+                break
+            except RateLimitError as e:
+                if 'You exceeded your current quota' in str(e._message):
+                    raise e
+                sleep(60)
+                continue
+            except Exception as e:
+                print(f'Caught exeception {str(e)}, continuing')
```

Comment:
we recently added proper logging to foundry, so please switch all print statements to use a logger